src/mds/Server.cc

```diff
@@ -8240,10 +8240,10 @@ void Server::handle_client_rename(MDRequestRef& mdr)
     if (!check_access(mdr, destdn->get_dir()->get_inode(), MAY_WRITE))
       return;
 
-    if (!check_fragment_space(mdr, destdn->get_dir()))
+    if (!linkmerge && !check_fragment_space(mdr, destdn->get_dir()))
```

Comment:
I'm a bit confused about what `linkmerge` signifies. It is set when source and dest refer to the same inode. Then, there is a check later in this function that asserts if the source inode is a stray inode when `linkmerge` is set, which looks to me that this flag signifies if the current operation is a stray inode reintegration.

If that's the case, should this be::

```
if (linkmerge && ....)
```

Reply:
> I'm a bit confused about what `linkmerge` signifies. It is set when source and dest refer to the same inode. Then, there is a check later in this function that asserts if the source inode is a stray inode when `linkmerge` is set, which looks to me that this flag signifies if the current operation is a stray inode reintegration.

Yes. Keep in mind that `rename(3)` syscall does not perform any action (surprisingly!) when `src` and `dest` are hardlinks to the same file. See also:

> If the old argument and the new argument resolve to the same existing file, rename() shall return successfully and perform no other action.

From: https://pubs.opengroup.org/onlinepubs/9699919799/

The MDS does not actually test this for this specific case -- it relies on the client to catch that and return without involving the MDS.

For this reason, the MDS assumes this is a stray reintegration whenever it sees a rename operation that does this.

> If that's the case, should this be::

No, we do not want to check fragment space if it's a `linkmerge`.

Reply:
> No, we do not want to check fragment space if it's a `linkmerge`.

doh! obviously. I incorrectly interpreted the condition ;P